File: csharp/tsptw-1-basic.cs
using System.Collections.Generic;
using System;
// this example illustrates how to use the api to solve a simple tsp with time windows.
// time windows are randomly generated so there's a change not all stops can be serviced
class tsptw1basic : IRunner
{
  public tsptw1basic(List<dataRow> data, string configFile = "../config.json")
  {
    this.configFile = configFile;
    this.data = data;
  }

  public void Run()
  {
    var api = new ApiHelper<Tsptw.SolveRequest, Tsptw.SolutionResponse>("tsptw-kcxbievqo879", configFile);
    // so here we're going to build the model 

    // create a solve request
    Tsptw.SolveRequest sr = new Tsptw.SolveRequest();

    sr.Model = new Tsptw.Tsp(); // initialise the model container

    Random rand = new Random();
    // add points to the tsp model
    foreach (var d in data)
    {
      // lets randomly create a window here.
      double rupper = 2500;
      double ws = rand.NextDouble() * rupper;
      double we = ws + rupper; //we don't accept backwards windows, so we'll just set these to some positive width upper amount.
      sr.Model.Points.Add(new Tsptw.Geocode
      {
        Id = d.id,
        X = d.X,
        Y = d.Y,
        windowStart = (float)ws,
        windowEnd = (float)we,
      });
    }

    // configure the distance metric (although road network is the default)
    sr.Model.Distancetype = Tsptw.Tsp.eDistanceType.RoadNetwork;

    string requestId = api.Post(sr); // send the model to the api
    Solution = api.Get(requestId);   // get the response (which it typed, so that's cool)

    // lets pull out the total distance
    lineString ls = new lineString();
    double totalDistance = 0.0;
    foreach (var e in Solution.Edges)
    {
      totalDistance += e.Distance;
      foreach (var g in e.Geometries)
      {
        ls.Add(new double[2] { g.X, g.Y });
      }
    }
    Console.WriteLine(string.Format("Total distance: {0:0.00} km\t Stops: " + Solution.Tours.Count, totalDistance));
    int maxchar = 0;
    foreach (var d in data)
    {
      maxchar = Math.Max(d.id.Length, maxchar); // so that the table displays nicely in the console :-) 
    }
    string formatLine = "|{0,-" + maxchar + "}|{1,11}|";
    Console.WriteLine(String.Format(formatLine, "Location", "ArrivalTime"));
    for (int i = 0; i < Solution.arrivalTimes.Length; i++)
    {
      Console.WriteLine(String.Format(formatLine, Solution.Tours[i], Solution.arrivalTimes[i]));
    }
    // just for fun - you wouldn't ever plot things this way - but it is nice to "see" the line segments
    Console.WriteLine("Approximate Tour Map:");
    new consolePlot(new List<lineString> { ls });

    return;
  }

  public Tsptw.SolutionResponse Solution { get; set; }

  private string configFile;

  private List<dataRow> data;
}

File: R/tsptw-1-basic.R
# TSPTW - test script
# devtools::load_all()
library(iceR)

# create an api-helper object with the model type you'd like to solve.
api <- new("apiHelper", modelType = 'tsptw-kcxbievqo879', configFile = '../config.json')

# create your model
sr <- new (TSPTW.SolveRequest)
sr$model <- new (TSPTW.TSP)

# add some data to the model
data <- read.csv('../sample_data/publist.csv')[1:10,]
# there are many ways to build protobuf. This is one simple way using old-school R functions.
rupper <- 2500
data <- data[sample(1:nrow(data), size = nrow(data), replace = F),]
# just randomly shuffle the selected points. We're making somewhat random windows below
# and if you run the script again you'll want to see a different result

data$WindowStart <- runif(nrow(data), 0, rupper)
                      #this creates some random time windows, with starts in [0,2500] and ends in
                      # [0,5000]. These are hard windows that have to be respected in terms of arrival
                      # so a "vehicle" will wait for the window to start if it arrives early at a location.

data$WindowEnd <- data$WindowStart + rupper
# we don't accept backwards windows, so we'll just set these to some positive width upper amount.

sr$model$points <- apply(X = data,
                         MARGIN = 1,
                         FUN = function(i){
                           g <- new (TSPTW.Geocode)
                           g$id <- as.character(i['id'])
                           g$x <- as.numeric(i['X'])
                           g$y <- as.numeric(i['Y'])
                           g$windowStart <- as.numeric(i['WindowStart'])
                           g$windowEnd <- as.numeric(i['WindowEnd'])
                           return(g)
                         })
sr$model$points[[1]]$windowStart <- sr$model$points[[1]]$windowEnd <- c()
sr$model$distancetype <- 1      # set the distance type to use the road-network

# if you'd like to view the model in plain-text format simply
sr$model$toString() %>% cat

requestID <- api %>% postSolveRequest(sr)         # submit the model to the api

resp <- api %>% getResponse(requestID)            # retrieve the model response

# a quick note. The objective with a tsp-tw is to minimise the total travel time.
# this means that a slightly longer route might be taken in terms of km-distance in order to
# effectively use the available time. Helps if you're working with real-world data for this example.

tab <- resp %>% tabulate(sr)                    # tabulate the data in native-R format.
tab$edges %>% arrange(sequence)
# What can happen now, which doesn't with a tsp, is that we may end up with unsatisfiable
# windows. In other words, try to execute all these orders, but the windows don't permit it
# to be done within the constraints. In this situation, we do as many of the stops as possible.
# Just randomise the data again if this happens.
resp$arrivalTimes

tab$nodes %>% head                              # just inspect the head of the output data
tab$edges %>% head                              # just inspect the head of the output data

resp %>% plotResponse(sr)                       # plot the data using ggplot. always looks nice.
resp %>% plotResponseLeaflet(sr)                # plot the data using leaflet (assuming it's mappable)

# because of the time windows, some stops may be ommitted (infeasible) and the sequence is probably
# not going to be the shortest in distance, but will be the shortest in time.
